src/storage-types/src/connections.rs

```diff
@@ -1262,6 +1266,239 @@ impl RustType<ProtoTunnel> for Tunnel<InlinedConnection> {
+            opts = opts.pass(Some(password));
+        }
+
+        // TODO(roshan): Implement SSL mode support
+        if self.tls_mode != MySqlSslMode::Disabled {
+            return Err(anyhow!("MySQL TLS modes are not yet supported"));
+        };
+
+        // TODO(roshan): Implement Root TLS Cert support
+        if self.tls_root_cert.is_some() {
+            return Err(anyhow!("MySQL TLS Certs are not yet supported"));
+        }
+
+        // TODO(roshan): Implement SSH Tunnels, AWS Privatelink
+        let tunnel = match &self.tunnel {
+            Tunnel::Direct => mz_mysql_util::TunnelConfig::Direct,
+            _ => {
+                return Err(anyhow!("MySQL Tunnels are not yet supported"));
+            }
+        };
+
+        Ok(mz_mysql_util::Config::new(opts.into(), tunnel))
+    }
+
+    async fn validate(
+        &self,
+        _id: GlobalId,
+        storage_configuration: &StorageConfiguration,
+    ) -> Result<(), anyhow::Error> {
+        let config = self
+            .config(
+                &*storage_configuration.connection_context.secrets_reader,
+                storage_configuration,
+            )
+            .await?;
+        let conn = config
+            .connect(
+                "connection validation",
+                &storage_configuration.connection_context.ssh_tunnel_manager,
+            )
+            .await?;
+        conn.disconnect().await?;
+        Ok(())
+    }
+}
+
+impl RustType<ProtoMySqlConnection> for MySqlConnection {
+    fn into_proto(&self) -> ProtoMySqlConnection {
+        ProtoMySqlConnection {
+            host: self.host.into_proto(),
+            port: self.port.into_proto(),
+            database: self.database.into_proto(),
+            user: Some(self.user.into_proto()),
+            password: self.password.into_proto(),
+            tls_mode: match self.tls_mode {
+                MySqlSslMode::Disabled => ProtoMySqlSslMode::Disabled.into(),
+                MySqlSslMode::Required => ProtoMySqlSslMode::Required.into(),
+                MySqlSslMode::VerifyCa => ProtoMySqlSslMode::VerifyCa.into(),
+                MySqlSslMode::VerifyIdentity => ProtoMySqlSslMode::VerifyIdentity.into(),
+            },
+            tls_root_cert: self.tls_root_cert.into_proto(),
+            tls_identity: self.tls_identity.into_proto(),
+            tunnel: Some(self.tunnel.into_proto()),
+        }
+    }
+
+    fn from_proto(proto: ProtoMySqlConnection) -> Result<Self, TryFromProtoError> {
+        Ok(MySqlConnection {
+            host: proto.host,
+            port: proto.port.into_rust()?,
+            database: proto.database,
+            user: proto.user.into_rust_if_some("ProtoMySqlConnection::user")?,
+            password: proto.password.into_rust()?,
+            tunnel: proto
+                .tunnel
+                .into_rust_if_some("ProtoMySqlConnection::tunnel")?,
+            tls_mode: match ProtoMySqlSslMode::from_i32(proto.tls_mode) {
```

Comment:
Since this is a separate type it would be better to provide a standalone `impl RustType<ProtoMySqlSslMode> for MySqlSslMode` implementation and use `proto.tls_mode.into_rust_if_some("...")?,` here